### DISCUSSION TITLE
Should we skip Schema validation on load?

### DISCUSSION THREAD

**Discussion by sdumetz** [Smithsonian/dpo-voyager#317] [Ideas]
I was doing a bit of performance analysis and tests show that initial schema validation by [DocumentValidator](https://github.com/Smithsonian/dpo-voyager/blob/master/source/client/io/DocumentValidator.ts) can get quite expensive.

On a medium-sized scene file ([this one](https://ethesaurus.uliege.be/ui/scenes/encensoir/view)), it adds 150 to 200ms to initial paint on a good laptop. Furthermore, it seems CPU-bound so it might be a lot worse on mobile devices (didn't test how bad it is though).

![flamegraph](https://github.com/user-attachments/assets/a5e79af7-28b6-42df-ad32-cfa429a038f4)

Do we really need schema validation?

 - It can't replace server-side validation if we **want** to ensure schema compliance.
 - If the document is non-compliant, we should try to display what we can anyways (think: backward compatibility when we add things to the schema for which lack of support might not be fatal )
 - Schema validation is used only there and adds ~200kb to the minified JS bundle

I'm not saying schema validation has no use, I understand its value, especially if Voyager is used on a simple webDAV backend that have no ability to perform said validation itself. But the price to pay with the current implementation seems too high. The solutions I see to improve on this are:

 1. drop client-side schema validation entirely (or only from production code)
 2. defer document validation to after initial load. We can always show a popup later to tell the user we expect things to break.
 3. precompile the schema, as advised in the [managing-schemas](https://ajv.js.org/guide/managing-schemas.html) guide

**1.** might be reasonable as it only prevents **trying to load** a bad document and there might not be a case where this is expected behaviour?

If that's not desirable, I can test if **3.**'s benefits would justify the additional tooling.

**Comment by gjcope** [comment]
@sdumetz Sorry, totally missed this somehow. My concern is that I don't think we do a good job of handling non-compliant documents in a number of cases and if (when) things break, we would really need to be able to point the user to the scene file being the issue. It's a valid point though as I don't think many users are currently hand-crafting scenes so non-compliance is probably an edge case. Hard to say though.

**Comment by sdumetz** [comment]
>  I don't think we do a good job of handling non-compliant documents

We don't do a particularly good job of handling compliant documents that might break things too. Even if we did, the scene file's format is not designed in a way that would allow schema compliance to mean anything.

And why a non-compliant documents that was generated from the UI  (https://github.com/Smithsonian/dpo-voyager/pull/338) and wouldn't break anything if it was allowed to run would completely bricks the scene, preventing the user from fixing it in the interface instead of hand-editing the JSON is VERY hard to explain to users.

We are and-crafting (or script-crafting) scenes a lot. But I don't think many others do. When we do, server-side scene validation is a better way of catching errors (we have quite a fancy script for this that does a deep merge on a parsed scene tree to allow some multi-user changes).

A marge proportion of errors when hand-editing scenes are off-by-one (or similar) errors in array indices, which no schema is ever going to catch (but tree-parsing will).


Of course, schema validation isn't without merits. Maybe I can move it to happen out of the critical path (to not slow down first-render and be recoverable), possibly also only in story mode?

**Reply by gjcope** [reply]
>We don't do a particularly good job of handling compliant documents that might break things too. Even if we did, the scene file's format is not designed in a way that would allow schema compliance to mean anything.

I'm not sure I follow. Do you have an example of this? If a compliant document breaks something, it should be considered a bug and fixed. Or maybe I'm missing your point. The schema validation has helped me to save time fixing scene file issues from users a number of times by pointing to missing/extra attributes, malformed attributes, etc. 

> And why a non-compliant documents that was generated from the UI (https://github.com/Smithsonian/dpo-voyager/pull/338) and wouldn't break anything if it was allowed to run would completely bricks the scene, preventing the user from fixing it in the interface instead of hand-editing the JSON is VERY hard to explain to users.

Fair point, but I would consider that a bug which should be (and was) fixed. If we're not going to enforce schema compliance, then there is not a lot of point to validation, which I guess is potentially what you are advocating.

Very intrigued by what you are doing with scene merges and multi-user changes. This is something that we have gotten requests for quite a bit. Maybe another discussion of you are able to share?

>Of course, schema validation isn't without merits. Maybe I can move it to happen out of the critical path (to not slow down first-render and be recoverable), possibly also only in story mode?

Seems entirely reasonable to move it to Story mode.

**Comment by sdumetz** [comment]
Compliant documents can break the app in any number of ways. The easiest way is to just mismatch any of the array-indexed pointers (like: `meta: 0` or `children: [1,2,3]`. If you insert a child but don't update all pointer references, the schema still validates (and there is no way to update it not to) but the scene won't work. Tour step indices are the same story but even harder to get right. Basically once you begin add/rm-ing nodes there is NO WAY you're not breaking the scene unless the script does it for you.

We just can't achieve "compliance means it runs" or anythign close without completely redesigning the document with strict schema validation in mind. This can be done (eg: OpenAPI), but we probably should not.

On the other hand, what bothers me is not the _existence_ of bugs caused by a schema / implementation mismatch: That's normal and not too frequent. But crashing the UI because of it is, I think, a problem.

So what I am advocating is to keep the schema validation, but move it so it doesn't make things worse when it's wrong (don't crash the doc parser), and (maybe?) only run it _where_ we can fix things: in the Story editor.

I'll start a discussion for the multi-user code. It's an experiment we have been running for some time now with mixed results. 99% of it is server-side so I you haven't seen much of it, but it's also open-source so I can easily show it.

**Comment by gjcope** [comment]
> Compliant documents can break the app in any number of ways. The easiest way is to just mismatch any of the array-indexed pointers (like: meta: 0 or children: [1,2,3]. If you insert a child but don't update all pointer references, the schema still validates (and there is no way to update it not to) but the scene won't work. Tour step indices are the same story but even harder to get right. Basically once you begin add/rm-ing nodes there is NO WAY you're not breaking the scene unless the script does it for you.

I get your point, but I think there's a level of reasonableness that has to be taken into account. I think there's a difference between the scene 'loading' and 'working'. I think it's reasonable to expect that a compliant document would load, but not necessarily work. If there's such a way to make a compliant document not load, I think that's a bug that needs to be fixed. Also keeping in mind that most of this is predicated on hand-editing the scene files. One of our major goals with Story is to make that unnecessary (I know we are not there yet). If there are major gaps on that end as well, it would be great to start a discussion/proposal.

> We just can't achieve "compliance means it runs" or anythign close without completely redesigning the document with strict schema validation in mind. This can be done (eg: OpenAPI), but we probably should not.

Again point taken, but I think we're just looking at it slightly differently. I think a compliant document should not crash the app and I think that's a reasonable expectation.

> So what I am advocating is to keep the schema validation, but move it so it doesn't make things worse when it's wrong (don't crash the doc parser), and (maybe?) only run it where we can fix things: in the Story editor.

Yes, I think running it only in Story makes sense if you are seeing significant performance issues. I still don't agree with it being 'wrong' per se, it's just only validating the syntactic part of it. The logical part of it (array indices pointing to the right places, etc.) is a different story.

> I'll start a discussion for the multi-user code. It's an experiment we have been running for some time now with mixed results. 99% of it is server-side so I you haven't seen much of it, but it's also open-source so I can easily show it.

Thanks so much, greatly appreciated!
